feat: documentation note formating (#1364)
AmadeusITGroup/otter: docs/localization/LOCALIZATION.md

class AppModule {}
```

> > [!INFO]

Reviewer: seems not working

class AppModule {}
```

> [!NOTE]
> The token `MESSAGE_FORMAT_CONFIG` implements the `LazyMessageFormatConfig` interface from `@o3r/localization`.

> [!TIP]
> The full documentation about MessageFormat configuration is available on <https://github.com/lephyrus/ngx-translate-messageformat-compiler>.

### How to localize plural expression

For pluralizations, we are using *TranslateMessageFormatCompiler*, coming from [ngx-translate-messageformat-compiler](https://www.npmjs.com/package/ngx-translate-messageformat-compiler) package, which is a compiler for *ngx-translate* that uses *messageformat.js* to compile translations using ICU syntax for handling pluralization and gender.
[ICU Message Format](http://userguide.icu-project.org/formatparse/messages) is a standardized syntax for dealing with the translation of user-visible strings into various languages that may have different requirements for the correct declension of words (e.g. according to number, gender, case) - or to simplify: pluralization.

> Simple pluralization rules like *0, 1 or other* fits well for English but may not fit for many other languages (Eastern Europe languages, Asiatic languages) where pluralization rules are much more complex. If this does not fit your requirement we recommend to reformulate your text, so that you do not need to use pluralization. Example: instead of saying 'You have added 2 baggages' you may want to say 'Pieces of baggage: 2' which should be fine for most languages no matter which number is considered to be plural.

Reviewer 2: ```suggestion
Simple pluralization rules like *0, 1 or other* fit well for English but may not fit for many other languages (Eastern Europe languages, Asian languages) where pluralization rules are much more complex. If this does not fit your requirement we recommend to reformulate your text, so that you do not need to use pluralization. Example: instead of saying 'You have added 2 baggages' you may want to say 'Pieces of baggage: 2' which should be fine for most languages no matter which number is considered to be plural.
```

Reviewer 3: ```suggestion
Simple pluralization rules like *0, 1 or other* work well in English but not necessarily in many other languages (Eastern European languages, Asian languages) where pluralization rules are much more complex. If this does not suit your requirements, we recommend reformulating your text so that you do not need to use the plural. Example: instead of saying 'You have added 2 pieces of luggage' you may want to say 'Pieces of luggage: 2', which should suit most languages, regardless of the number considered as plural.
```

#### Integration with ngx-translate

> You need to configure *TranslateModule* for it to use *TranslateMessageFormatCompiler* as compiler. We will use *TranslateMessageFormatLazyCompiler* which is an otter extension of the base compiler. See *Lazy Compiler for ICU Translation support* above chapter for details.

Reviewer 2: ```suggestion
You need to configure *TranslateModule* so that it uses *TranslateMessageFormatCompiler*. We will use *TranslateMessageFormatLazyCompiler* which is an otter extension of the base compiler. See *Lazy Compiler for ICU Translation support* above chapter for details.
```

Reviewer 2: To be honest at this point, I am not sure which compiler we use between TranslateMessageFormatCompiler and TranslateMessageFormatLazyCompiler

</span> {{translations.nbOfErrors | o3rTranslate: {count: countMessages} }}
...
```

> The value of *translations.nbOfErrors* is the translation key 'o3r-list-inline-messages-pres.nbOfErrors'. The next step translates the key passing some parameters to translate pipe.

Reviewer 2: ```suggestion
The value of *translations.nbOfErrors* is the translation key 'o3r-list-inline-messages-pres.nbOfErrors'. 
Next, the key and some parameters (count in this case) are forwared to the translate pipe.
```

  <li>{{ translations.people | o3rTranslate: { how: 'affectionate' } }}</li>
</ul>
```

> Note again that *translations.people* matches *global.people* key **Output**

Reviewer 2: ```suggestion
Note again that *translations.people* matches *global.people* key 

**Output**
```

## Runtime: toggle translation on and off

In order to be able to more easily identify which key corresponds to a given text, the ``LocalizationService`` exposes a function ``toggleShowKeys()`` that can be called in order to deactivate or reactivate the translation mechanism at **runtime**.
While deactivated, the ``translate`` **pipe** and **directive** will output the translation keys instead of their resolved values.

> > **Important**: this mechanism only applies to the pipe and directive exported by Otter's ``LocalizationModule``. The original ones from ``ngx-translate`` do not support it.

Reviewer 4: Why not 
```suggestion
> [!IMPORTANT]
> This mechanism only applies to the pipe and directive exported by Otter's ``LocalizationModule``. The original ones from ``ngx-translate`` do not support it.
```

## Runtime: toggle translation on and off

In order to be able to more easily identify which key corresponds to a given text, the ``LocalizationService`` exposes a function ``toggleShowKeys()`` that can be called in order to deactivate or reactivate the translation mechanism at **runtime**.
While deactivated, the ``translate`` **pipe** and **directive** will output the translation keys instead of their resolved values.

> [!IMPORTANT]
> This mechanism only applies to the pipe and directive exported by Otter's ``LocalizationModule``. The original ones from ``ngx-translate`` do not support it.

First, this mechanism has to be activated via the ``LocalizationConfiguration`` that you can provide in your ``ApplicationModule``.

> This is mainly for performances reason: the way it works is it adds a new subscription to every ``translate`` pipe and directive in order to know when translations are turned on or off.

Reviewer 2: ```suggestion
This is mainly for performances reason as this adds a new subscription to every ``translate`` pipe and directive in order to know when translations are turned on or off.
```

> Not enabling it allows to avoid all those subscriptions, and should be the baseline for a production environment.

Reviewer 2: ```suggestion
Not enabling it prevents all those subscriptions, and should be the baseline for a production environment.
```

> You can enable debug mode by setting debugMode property of LocalizationConfiguration (by default the value of debugMode is false)

Reviewer 2: ```suggestion
You can enable debug mode by enabling the debugMode property of LocalizationConfiguration (by default the value of debugMode is false)
```

> By doing this all your translations will be prefixed by the corresponding localization key. That way you can easily map the text to the key.

Reviewer 2: ```suggestion
By doing this, all your translations will be prefixed with by the corresponding localization key. That way you can easily map the text to the key.
```

}
```

> [!NOTE]

> > get more details on [dev tools session](../dev-tools/chrome-devtools.md)

Reviewer 2: ```suggestion
> Get more details on [dev tools session](../dev-tools/chrome-devtools.md)
```

```

# Intelligent fallback support

> The fallback hierarchy has been added to the localization service which explained in detail below. As the last option, the default fallback language picked as usual.

Reviewer 2: ```suggestion
The localization service supports a fallback strategy described in the details below. 
```

## Scenario 1: Fallback based on **fallbackLocalesMap**, **supportedLocales** language code

> In case **fallbackLocalesMap** is provided and the targeted translation language is unavailable in supported locales list, The priority goes to **fallbackLocalesMap**, to see if targeted translation language can be mapped with the fallback map configured. The second priority goes to **supportedLocales**, to map targeted translation language to the first closest possible language, the locale can be different. If none match, it will fall back to the **default language**.

Reviewer 2: ```suggestion
In case **fallbackLocalesMap** is provided and the targeted translation language is unavailable in supported locales list, the priority goes to **fallbackLocalesMap**, to see if the targeted translation language can be mapped to the fallback map. Then, the second priority goes to **supportedLocales**, to map the targeted translation language to the first closest possible language -- for instance a different locale of the requested language (en-GB to en-US). 

If none matches, it will fallback to the **default language**.
```

bn-BD **fallbacks to** ar-AR, as it is the default fallback.

## Scenario 2: Fallback based on **supportedLocales** language code

> In case if **fallbackLocalesMap** is not provided and the targeted translation language is unavailable is supported locales list, The targeted translation language will be matched with supported locales **language** to find out the first nearest match, the **locale** can be different. If none match, it will fall back to the **default language**.

Reviewer 2: ```suggestion
If **fallbackLocalesMap** is not provided and the targeted translation language is not part of the supported locale list, the targeted translation language will be matched to the supported locales' **languages** to find out the first nearest one. Again, only the language code will be matched and the locale will be ignored. 

If none match, it will fallback to the **default language**.
```
